privateSBOMExchange/src/petra/lib/models/__init__.py

```diff
@@ -61,3 +65,166 @@ def tree_from_nodes(nodes, values, root):
     new_tree.root=root
 
     return new_tree
+
+#this function is just to try things, should be replaced by the real cpabe
+def cpabe(data):
+    if not isinstance(data, bytes):
+        return data.encode()
+    else:
+        return data
+        
+class Node:
+    """Base class for a node in the Merkle tree."""
+    def accept(self, visitor):
+        raise NotImplementedError("Must implement accept method.")
+
+class Leaf(Node):
+    """Represents a leaf node containing binary data."""
+    def __init__(self, data):
+        self.data = data  # Store data as bytes
+
+    def accept(self, visitor):
+        return visitor.visit_leaf(self)
+
+class InternalNode(Node):
+    """Represents an internal node that can have multiple children."""
+    def __init__(self, package_name, children):
+        self.data = package_name  # Store package name as data
+        self.children = children
```

Comment:
What's the data structure of children? This also seems to assume that every internal node is a package, but it may actually be the root of another SBOM tree, right?

Reply:
the data structure of children is Leaf -> fields in the sbom
Internal nodes -> pkgs
root is the sbom
.. 
the name might be misleading in a way
but the internal nodes are not meant to be nodes that are not leaves nor roots in a tree, internal nodes are package nodes that we don't have sboms for
the root node represent the root of this sbom or the root of any dependency that has sbom

the root children can of type: root(dependency has sbom)/internal node(pkg)/leaf(field data)

Reply:
do you think while we visit the tree using the Merkle visitor, we should replace the raw data by the hashed merkle data?

Reply:
By "data structure of children" I had meant how the references to the lead nodes are being stored in the internal nodes. That is, is `self.children` an array, a linked list, a dict etc.? This isn't clear from the code, and the assumption should be documented, because it could affect how the Merkle visitor hashes things, it could affect search performance etc.

> internal nodes are not meant to be nodes that are not leaves nor roots in a tree, internal nodes are package nodes that we don't have sboms for

If this is the case, my recommendation is to use clearer names for the different types, which actually indicate their purpose. If we're to have package nodes and they have unique properties, let's call them that :)

> the root children can of type: root(dependency has sbom)/internal node(pkg)/leaf(field data)

> the root node represent the root of this sbom or the root of any dependency that has sbom

I agree with this description, and if I'm being really nitpicky, by definition this should no longer be called a root node, but something like a "SBOM node" since it may or may not be at the root of a tree. Our main goal should be to express the unique aspects of a node that represents a field vs a package (with no SBOM) vs an SBOM.

> do you think while we visit the tree using the Merkle visitor, we should replace the raw data by the hashed merkle data?

No, we shouldn't be replacing the contents of a node when we hash it, we should be storing the hash in a separate field.

Reply:
@marcelamelara Thank you for the feedback! I addressed your feedback in the last commit, please take a look.
 how can we address this goal in the code " express the unique aspects of a node that represents a field vs a package (with no SBOM) vs an SBOM."?

Reply:
> how can we address this goal in the code " express the unique aspects of a node that represents a field vs a package (with no SBOM) vs an SBOM."?

I think this relates to our conversation on Slack about the fine-grained data fields. The way I would do this is to identify the important properties that describe a field, a package (with no SBOM), and an SBOM and define these as separate variables in the respective node implementations.